flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ManagedTableListener.java

```diff
@@ -0,0 +1,150 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.table.api.TableException;
+import org.apache.flink.table.descriptors.ConnectorDescriptorValidator;
+import org.apache.flink.table.factories.DynamicTableFactory;
+import org.apache.flink.table.factories.FactoryUtil;
+import org.apache.flink.util.StringUtils;
+
+import javax.annotation.Nullable;
+
+import java.util.Map;
+
+import static org.apache.flink.table.catalog.CatalogBaseTable.TableKind.MANAGED;
+import static org.apache.flink.table.factories.ManagedTableFactory.discoverManagedTableFactory;
+
+/** The listener for managed table operations. */
+@Internal
+public class ManagedTableListener {
+
+    private final ClassLoader classLoader;
+
+    private final ReadableConfig config;
+
+    public ManagedTableListener(ClassLoader classLoader, ReadableConfig config) {
+        this.classLoader = classLoader;
+        this.config = config;
+    }
+
+    /** Notify for creating managed table. */
+    public ResolvedCatalogBaseTable<?> notifyTableCreation(
+            @Nullable Catalog catalog,
+            ObjectIdentifier identifier,
+            ResolvedCatalogBaseTable<?> table,
+            boolean isTemporary,
+            boolean ignoreIfExists) {
+        if (isNewlyManagedTable(catalog, table)) {
+            ResolvedCatalogTable managedTable = createManagedTable(identifier, table, isTemporary);
+            discoverManagedTableFactory(classLoader)
+                    .onCreateTable(
+                            createTableFactoryContext(identifier, managedTable, isTemporary),
+                            ignoreIfExists);
+            return managedTable;
+        }
+        return table;
+    }
+
+    /** Notify for dropping managed table. */
+    public void notifyTableDrop(
+            ObjectIdentifier identifier,
+            ResolvedCatalogBaseTable<?> table,
+            boolean isTemporary,
+            boolean ignoreIfNotExists) {
+        if (table.getTableKind() == MANAGED) {
+            discoverManagedTableFactory(classLoader)
+                    .onDropTable(
+                            createTableFactoryContext(
+                                    identifier, (ResolvedCatalogTable) table, isTemporary),
+                            ignoreIfNotExists);
+        }
+    }
+
+    private boolean isNewlyManagedTable(
+            @Nullable Catalog catalog, ResolvedCatalogBaseTable<?> table) {
+        if (catalog == null || !catalog.supportsManagedTable()) {
+            // catalog not support managed table
+            return false;
+        }
+
+        if (table.getTableKind() == CatalogBaseTable.TableKind.VIEW) {
+            // view is not managed table
+            return false;
+        }
+
+        if (!StringUtils.isNullOrWhitespaceOnly(
```

Comment:
we call `table.getOptions()` very often here. how about calling it once in a try-catch and than just check for the remaining cases?